lesson4/hometask4_mkolesnik.rb

```diff
@@ -0,0 +1,42 @@
+class LazyHash < Hash
+
+  def method_missing(method_name, *args, &block)
+    if !self[*args] && self.keys.include?(method_name) && lazy?(method_name)
+      if block_given?
+        self[*args] = yield self[method_name].call(*args, &block)
+      else
+        self[*args] = self[method_name].call(*args, &block)
+      end
+    elsif self.keys.include?(*args)
+      block_given? ? (yield self[*args]) : self[*args]
+    else
+      super
+    end
+  end
+
+  def respond_to_missing?(method_name, include_private=false)
+    self.methods.include?(method_name) || super
+  end
+
+  def lazy?(*args)
+    key, value = args
+    self[key].is_a?(Proc) ^ self.keys.include?(value)
+  end
+end
+
+
+h = LazyHash.new
+h[:factorial] = proc { |num| (1..num).reduce(:*) }
+
+p h.lazy?(:factorial) # true
+p h.lazy?(:factorial, 5) # true
+p h[5] # nil
+p h.factorial(5) # 120
+p h[5] # 120
+p h.lazy?(:factorial, 5) # false
+
+p h[:factorial]
+p h.factorial(1000) { |num| num.to_s.reverse }
```

Comment:
I don't think that output:

> true
> true
> nil
> 120
> 120
> false
> #<Proc:0x007fc42d049cc0@temp.rb:29>
> "000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000274357720079014939777538389339501096351400909614342172426007309558065032425273699690048180853531032928061587147595404776360104285371247452490899847278149508651754993451281046015820249680988021903920973062164615442107178499054121659082628959909693368848357261829869780528894965036375702022443257567685757842228247691772455368983795200590712605380931123468987206468962882872334251370990965446770384351092496993150510828443321410851025811434182496049812733578946629135824731079341353946681420521321192283756938416246883273918692915809227897069421646558121783488860129146109548922171120113819927988118040068710798391653207997160584625665810688109933409567122222094998952046644013767041009418383729164679886568384032835783773268548224517269643314110477947822456955027859315565794372394631320929026091539074252588128204410542756564834234420480097301375286495064797136443602422725496534103006855831216081889390630255651131559564972051683887068138216999671066763340662399528443843461802256988871858193187570708272286256859967625660854622861723854521095284869798096160339682479341622403384071562045522482615157025387851062711609687630238564012716181351423420992075559203826647185659246299547163781210282414444591500889982226472130457309119977909861971267616176342362815469398988087351061802934280432119265796476641769268672605880674179838322069781298260164145568322752545491411871393910887709015129571869301207116133741535888510002526253430948629167721703516738731057082637993054931206655553327101913359959489784244583109891804353567061549442646976988124827748574616735059009206579807640489645638516355118611681333128080275238665122243909914579469411507829953698442167290312018462182893196351420563461071868536487407779728143720151386162000802087805214131632624844231961767023985874520213843178576906747861612403753108679630116855132182304448293134803792878090829102395685122425435973580631647137955473838768621808942885917358341973702192028866693797766116675646824355743943474714262411904755570364859323581315234537559538988236674210779948746387528540181911733784688734788393650192694646490872814106340794264579478021672599059505977915237881727726328819957370780405424799095449966928323109855808499278666136850691791016889974008404969684802440295020926893215839924019997345236417012468473917589300329334207345377390770062783204"
> "000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000274357720079014939777538389339501096351400909614342172426007309558065032425273699690048180853531032928061587147595404776360104285371247452490899847278149508651754993451281046015820249680988021903920973062164615442107178499054121659082628959909693368848357261829869780528894965036375702022443257567685757842228247691772455368983795200590712605380931123468987206468962882872334251370990965446770384351092496993150510828443321410851025811434182496049812733578946629135824731079341353946681420521321192283756938416246883273918692915809227897069421646558121783488860129146109548922171120113819927988118040068710798391653207997160584625665810688109933409567122222094998952046644013767041009418383729164679886568384032835783773268548224517269643314110477947822456955027859315565794372394631320929026091539074252588128204410542756564834234420480097301375286495064797136443602422725496534103006855831216081889390630255651131559564972051683887068138216999671066763340662399528443843461802256988871858193187570708272286256859967625660854622861723854521095284869798096160339682479341622403384071562045522482615157025387851062711609687630238564012716181351423420992075559203826647185659246299547163781210282414444591500889982226472130457309119977909861971267616176342362815469398988087351061802934280432119265796476641769268672605880674179838322069781298260164145568322752545491411871393910887709015129571869301207116133741535888510002526253430948629167721703516738731057082637993054931206655553327101913359959489784244583109891804353567061549442646976988124827748574616735059009206579807640489645638516355118611681333128080275238665122243909914579469411507829953698442167290312018462182893196351420563461071868536487407779728143720151386162000802087805214131632624844231961767023985874520213843178576906747861612403753108679630116855132182304448293134803792878090829102395685122425435973580631647137955473838768621808942885917358341973702192028866693797766116675646824355743943474714262411904755570364859323581315234537559538988236674210779948746387528540181911733784688734788393650192694646490872814106340794264579478021672599059505977915237881727726328819957370780405424799095449966928323109855808499278666136850691791016889974008404969684802440295020926893215839924019997345236417012468473917589300329334207345377390770062783204"
> "402387260077093773543702433923003985719374864210714632543799910429938512398629020592044208486969404800479988610197196058631666872994808558901323829669944590997424504087073759918823627727188732519779505950995276120874975462497043601418278094646496291056393887437886487337119181045825783647849977012476632889835955735432513185323958463075557409114262417474349347553428646576611667797396668820291207379143853719588249808126867838374559731746136085379534524221586593201928090878297308431392844403281231558611036976801357304216168747609675871348312025478589320767169132448426236131412508780208000261683151027341827977704784635868170164365024153691398281264810213092761244896359928705114964975419909342221566832572080821333186116811553615836546984046708975602900950537616475847728421889679646244945160765353408198901385442487984959953319101723355556602139450399736280750137837615307127761926849034352625200015888535147331611702103968175921510907788019393178114194545257223865541461062892187960223838971476088506276862967146674697562911234082439208160153780889893964518263243671616762179168909779911903754031274622289988005195444414282012187361745992642956581746628302955570299024324153181617210465832036786906117260158783520751516284225540265170483304226143974286933061690897968482590125458327168226458066526769958652682272807075781391858178889652208164348344825993266043367660176999612831860788386150279465955131156552036093988180612138558600301435694527224206344631797460594682573103790084024432438465657245014402821885252470935190620929023136493273497565513958720559654228749774011413346962715422845862377387538230483865688976461927383814900140767310446640259899490222221765904339901886018566526485061799702356193897017860040811889729918311021171229845901641921068884387121855646124960798722908519296819372388642614839657382291123125024186649353143970137428531926649875337218940694281434118520158014123344828015051399694290153483077644569099073152433278288269864602789864321139083506217095002597389863554277196742822248757586765752344220207573630569498825087968928162753848863396909959826280956121450994871701244516461260379029309120889086942028510640182154399457156805941872748998094254742173582401063677404595741785160829230135358081840096996372524230560855903700624271243416909004153690105933983835777939410970027753472000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
> "402387260077093773543702433923003985719374864210714632543799910429938512398629020592044208486969404800479988610197196058631666872994808558901323829669944590997424504087073759918823627727188732519779505950995276120874975462497043601418278094646496291056393887437886487337119181045825783647849977012476632889835955735432513185323958463075557409114262417474349347553428646576611667797396668820291207379143853719588249808126867838374559731746136085379534524221586593201928090878297308431392844403281231558611036976801357304216168747609675871348312025478589320767169132448426236131412508780208000261683151027341827977704784635868170164365024153691398281264810213092761244896359928705114964975419909342221566832572080821333186116811553615836546984046708975602900950537616475847728421889679646244945160765353408198901385442487984959953319101723355556602139450399736280750137837615307127761926849034352625200015888535147331611702103968175921510907788019393178114194545257223865541461062892187960223838971476088506276862967146674697562911234082439208160153780889893964518263243671616762179168909779911903754031274622289988005195444414282012187361745992642956581746628302955570299024324153181617210465832036786906117260158783520751516284225540265170483304226143974286933061690897968482590125458327168226458066526769958652682272807075781391858178889652208164348344825993266043367660176999612831860788386150279465955131156552036093988180612138558600301435694527224206344631797460594682573103790084024432438465657245014402821885252470935190620929023136493273497565513958720559654228749774011413346962715422845862377387538230483865688976461927383814900140767310446640259899490222221765904339901886018566526485061799702356193897017860040811889729918311021171229845901641921068884387121855646124960798722908519296819372388642614839657382291123125024186649353143970137428531926649875337218940694281434118520158014123344828015051399694290153483077644569099073152433278288269864602789864321139083506217095002597389863554277196742822248757586765752344220207573630569498825087968928162753848863396909959826280956121450994871701244516461260379029309120889086942028510640182154399457156805941872748998094254742173582401063677404595741785160829230135358081840096996372524230560855903700624271243416909004153690105933983835777939410970027753472000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
> 

a good example to show expected behavior.